bcda/models/postgres/repository.go

```diff
@@ -0,0 +1,113 @@
+package postgres
+
+import (
+	"strconv"
+	"time"
+
+	"github.com/CMSgov/bcda-app/bcda/constants"
+
+	"github.com/CMSgov/bcda-app/bcda/models"
+	"github.com/jinzhu/gorm"
+)
+
+// Ensure Repository satisfies the interface
+var _ models.Repository = &Repository{}
+
+type Repository struct {
+	db *gorm.DB
+}
+
+func NewRepository(db *gorm.DB) *Repository {
+	return &Repository{db}
+}
+
+func (r *Repository) GetLatestCCLFFile(cmsID string, cclfNum int, importStatus string, lowerBound, upperBound time.Time) (*models.CCLFFile, error) {
+
+	const (
+		queryNoTime     = "aco_cms_id = ? AND cclf_num = ? AND import_status = ?"
+		queryLower      = queryNoTime + " AND timestamp >= ?"
+		queryUpper      = queryNoTime + " AND timestamp <= ?"
+		queryLowerUpper = queryNoTime + " AND timestamp >= ? AND timestamp <= ?"
+	)
+
+	var (
+		cclfFile models.CCLFFile
+		result   *gorm.DB
+	)
+	if lowerBound.IsZero() && upperBound.IsZero() {
+		result = r.db.Where(queryNoTime,
+			cmsID, cclfNum, constants.ImportComplete)
+	} else if !lowerBound.IsZero() && upperBound.IsZero() {
+		result = r.db.Where(queryLower,
+			cmsID, cclfNum, constants.ImportComplete,
+			lowerBound)
+	} else if lowerBound.IsZero() && !upperBound.IsZero() {
+		result = r.db.Where(queryUpper,
+			cmsID, cclfNum, constants.ImportComplete,
+			upperBound)
+	} else {
+		result = r.db.Where(queryLowerUpper,
+			cmsID, cclfNum, constants.ImportComplete,
+			lowerBound, upperBound)
+	}
+
+	result = result.Order("timestamp DESC").First(&cclfFile)
+	if result.RecordNotFound() {
+		return nil, nil
+	}
+
+	return &cclfFile, result.Error
+}
+
+func (r *Repository) GetCCLFBeneficiaryMBIs(cclfFileID uint) ([]string, error) {
+	var mbis []string
+
+	if err := r.db.Table("cclf_beneficiaries").Where("file_id = ?", cclfFileID).Pluck("mbi", &mbis).Error; err != nil {
+		return nil, err
+	}
+
+	return mbis, nil
+}
+
+func (r *Repository) GetCCLFBeneficiaries(cclfFileID uint, ignoredMBIs []string) ([]*models.CCLFBeneficiary, error) {
+
+	const (
+		// this is used to get unique ids for de-duplicating MBIs that are listed multiple times in the CCLF8 file
+		idQuery = "SELECT id FROM ( SELECT max(id) as id, mbi FROM cclf_beneficiaries WHERE file_id = ? GROUP BY mbi ) as id"
+	)
+	var beneficiaries []*models.CCLFBeneficiary
+
+	// NOTE: We changed the query that was being used for "old benes"
+	// By querying by IDs, we really should not need to also query by the corresponding MBIs as well
+	query := r.db.Where("id in (?)", r.db.Raw(idQuery, cclfFileID).SubQuery())
```

Comment:
The old version of this query used `mbi IN (oldMBIs) AND id IN (uniqueIDs)` for old benes and `mbi NOT IN (oldMBIs) and ids IN (uniqueIDs)`.

By querying for just `id IN (uniqueIDs)`, we'll roundtrip both old and new benes in a single query and filter them on the host.